File: 算法训练营/树/094/迭代法.py
#
# @lc app=leetcode.cn id=94 lang=python3
#
# [94] 二叉树的中序遍历
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def inorderTraversal(self, root: TreeNode) -> List[int]:
        stack = [root]
        records, ret = set(), []
        while stack:
            elem = stack.pop()
            if elem is None:
                continue
            if elem.left and elem.left not in records:
                stack.extend([elem, elem.left])
                continue
            ret.append(elem.val)
            records.add(elem)
            if elem.right:
                stack.append(elem.right)
        return ret
        
# @lc code=end

File: 算法训练营/03-数组链表跳表/283/迭代法.go
/*
 * @lc app=leetcode.cn id=283 lang=golang
 *
 * [283] 移动零
 */

// @lc code=start
func moveZeroes(nums []int)  {
	zero := 0 // 0 出现次数
	i := 0
	for ; i < len(nums); i++ {
		if nums[i] == 0 {
			zero++
			continue
		}
		// 不相等就交换
		if i != i-zero {
			nums[i-zero] = nums[i]
		}
	}
	// 末尾归 0
	for i := len(nums); zero > 0; zero-- {
		nums[i-zero] = 0
	}
}
// @lc code=end
